typing/typecore.ml

```diff
@@ -2746,25 +2757,38 @@ and type_expect_
         loc sexp.pexp_attributes env ty_expected_explained Nolabel caselist
   | Pexp_apply(sfunct, sargs) ->
       assert (sargs <> []);
-      begin_def (); (* one more level for non-returning functions *)
-      if !Clflags.principal then begin_def ();
-      let funct = type_exp env sfunct in
```

Comment:
Couldn't all the processing rather be done here.
I.e., if `sfunct` is a `Pexp_ident` then check whether it is `%apply` or `%revapply`.
I see no point in doing this recursively, as it will not match in the recursive call (there will be only one argument left).

Reply:
Good point!  I had in mind (but forgot to implement it) to rewrite `y |> f x` into `f x y`, and not `(f x) y`; and then it would make sense to allow `y |> (x |> f)`.  Do you think it would be better not to do that, and keep rewriting `y |> f x` to `(f x) y`?

Reply:
Actually, there are subtle points about the interactions of parentheses with optional arguments:
```ocaml
# let f ?(a = 0) () = a+1;;
val f : ?a:int -> unit -> int = <fun>
# f () ~a:3;;
- : int = 4
# (f ()) ~a:3;;
Error: This expression has type int
       This is not a function; it cannot be applied.
```
Namely, non-labeled arguments commute with optional arguments, but only inside the same multi-application. Otherwise we would not be able to discard non-applied optional arguments.
So merging arguments is not always neutral.
It is probably fine if you restrict it to the case where there are no labelled arguments passed after the merging point.

I'll try to think more about the implications.

Reply:
Indeed, here, we would only add one non-labeled argument (namely, the lhs of `|>`, or the rhs of `@@`).